File: Graph Theory Peaks/DFS/Q2. Get Watched Videos by Your Friends/Q2_GetWatchedVideosbyYourFriends.py
from collections import deque


class Solution(object):
    def watchedVideosByFriends(self, watchedVideos, friends, id, level):
        """
        :type watchedVideos: List[List[str]]
        :type friends: List[List[int]]
        :type id: int
        :type level: int
        :rtype: List[str]
        """

        visited = set()              # Set to track visited friends
        queue = deque()              # Queue for BFS traversal

        queue.append((id, 0))        # Start BFS from given person with level 0
        visited.add(id)              # Mark starting person as visited
        
        level_friends = []           # Store friends exactly at the given level
        
        # Perform BFS
        while queue:
            curr_id, curr_level = queue.popleft()   # Get current person and level

            # If we reached the required level
            if curr_level == level:
                level_friends.append(curr_id)       # Add this friend to the list

            # If current level is less than target level, continue exploring
            elif curr_level < level:
                for friend in friends[curr_id]:     # Check all friends of current person

                    # If friend not visited yet
                    if friend not in visited:
                        visited.add(friend)         # Mark friend as visited
                        queue.append((friend, curr_level + 1))  # Add friend with next level

        video_count = Counter()      # Dictionary to count video frequencies

        # Count videos watched by friends at the required level
        for friend in level_friends:
            for video in watchedVideos[friend]:
                video_count[video] += 1   # Increase count for this video
        
        # Sort videos:
        # 1. By increasing frequency
        # 2. If same frequency, by lexicographical order
        return sorted(video_count.keys(), key=lambda x: (video_count[x], x))

File: Graph Theory Peaks/DFS/Q2. Get Watched Videos by Your Friends/Q2_GetWatchedVideosbyYourFriends.java
class Solution {
    public List<String> watchedVideosByFriends(List<List<String>> watchedVideos, int[][] friends, int id, int level) {
        
        boolean[] visited = new boolean[friends.length]; // Track visited friends
        Queue<Integer> queue = new ArrayDeque<>();       // Queue for BFS traversal
        
        queue.offer(id);      // Start BFS from given person
        visited[id] = true;   // Mark starting person as visited
        
        // BFS until we reach the required friendship level
        while (!queue.isEmpty() && level-- > 0) {
            int size = queue.size();  // Number of nodes at current level
            
            while (size-- > 0) {
                int cur = queue.poll(); // Current friend
                
                // Explore all direct friends of current person
                for (int i : friends[cur]) {
                    if (!visited[i]) {
                        queue.offer(i);   // Add friend to queue
                        visited[i] = true; // Mark as visited
                    }
                }
            }
        }

        // If no friends exist at required level
        if (queue.isEmpty() || level > 0) return new ArrayList<>();

        Map<String, Integer> freq = new HashMap<>(); // Store video watch frequency

        // Collect videos watched by friends at target level
        for (int i : queue) {
            for (String video : watchedVideos.get(i)) {
                freq.put(video, freq.getOrDefault(video, 0) + 1); // Count frequency
            }
        }

        // Create result list from video names
        List<String> res = new ArrayList<>(freq.keySet());

        // Sort by frequency first, then lexicographically
        Collections.sort(res, (e1, e2) -> 
            freq.get(e1) == freq.get(e2) 
            ? e1.compareTo(e2)              // If freq same → sort by name
            : Integer.compare(freq.get(e1), freq.get(e2)) // Else sort by freq
        );

        return res; // Return sorted video list
    }
}